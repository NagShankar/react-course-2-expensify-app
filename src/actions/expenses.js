import uuid from 'uuid4';


//ADD_EXPENSE action creator
export const addExpense = ({description='', note='', amount=0, createdAt=0}={}) => {
    return {
        type:"ADD_EXPENSE",
        expense:{
            id:uuid(),
            description, //or description= description, or use shortand property
            note,
            amount,
            createdAt
        }
        
    }
}


//REMOVE_EXPENSE
export const removeExpense = ({id}={}) => {
    return {
        type:"REMOVE_EXPENSE",
        id //or id:id
        
    }
}



//EDIT_EXPENSE
export const editExpense = (id, updates) => {
    return {
        type:"EDIT_EXPENSE",
        id,
        updates
        
    }
}


//we are doing named export