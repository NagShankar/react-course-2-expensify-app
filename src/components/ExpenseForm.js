import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
//import "react-dates/lib/css/_datepicker.css"; //moved to app.js so other components can make use of it, just to avoid confusion of inporting it everywhere

//const now = moment();
//console.log(now.format('MMM Do, YYYY'))

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props); //now props will have the 'expense prop' sent from Edit Expense page to populate the fields to edit it
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount.toString() : "", //converting back to string while editing which we had earlier converted to number before submitting it to store
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(), //providing initial value i.e by default today's date
      calendarFocused: false,
      error: "",
    };
  }

  //WE CAN COMBINE DESCRIPTION AND NOTE IN ONE EVENT HANDLER LIKE BELOW
  //onValueChange = (e) => {
  //    const name=e.target.name;
  //    const value=e.target.value;
  //
  //    this.setState({
  //        [name]:value
  //    })
  //
  //}

  //   componentDidMount() {
  //     console.log(this.state);
  //   }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    //you cant assign e.target.value in setState directly, you cant do like this, this will throw an synthetic error, if you really want to assign target value directly in setState use e.persist() before doing that as suggested by that error
    //e.persist();
    //this.setState(()=>({note:e.target.value}))
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    //below matches empty string and regex with it
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (date) => {
    //, it activates a calendar again, it takes an argument(can be Date or Now and any name) in the form of moment and gives the back the date to reset the date
    //not letting user to clear date field, now setting onChange only when there's date already present there, thus forcing date to be present all the time and not letting user to chnage it at all, and change only if there's date
    //console.log(date.valueOf()); //new moment object is returned, and valueOf will convert time and date in milliseconds since Unix Epoch
    if (date) {
      this.setState(() => ({ createdAt: date }));
    }
  };
  onFocusChange = ({ focused }) => {
    //as given in official docs, destructuring the focused property, which have boolean value = TRUE on clicking its
    //console.log(focused);
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault(); //avoiding refresh if they press submit, which is browser default behaviour

    //validation to check if the value is present for fields like description and amount, only then make submit
    if (!this.state.description || !this.state.amount) {
      //alert("please enter value for desc and amount")
      //set error state to "Please provide description and amount"
      this.setState(() => ({
        error: "Please provide description and amount",
      }));
    } else {

      this.props.onSubmission({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10),
        createdAt: this.state.createdAt.valueOf(), //if valueOf not provided, then it displays moment object
      });
      //console.log('submitted');
        
        this.setState(() => ({
        //now emptying all fields after success
        description: "",
        note: "",
        amount: "",
        createdAt: moment(), //providing initial value
        calendarFocused: false,
        error: "",
      }));
        
        
        
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            name="description"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={
              1
            } /*change this number to see as many calendars you want */
            isOutsideRange={() =>
              false
            } /*enables past/finished dates as well, you can set range also, refer doc */
          />

          <textarea
            name="note"
            placeholder="Add a note for your expense(optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
