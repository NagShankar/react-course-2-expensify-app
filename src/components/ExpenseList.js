import React from "react";
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleFilters from './../selectors/expenses'

export const ExpenseList = (props) => {
    return(
     <div className="content-container">
        <div className="list-header">
           <div className="show-for-mobile">Expenses</div>
           <div className="show-for-desktop">Expense</div>
           <div className="show-for-desktop">Amount</div>
        </div>
        
      
       {/* {props.expenses.length}
        {props.filters.text} */}
            <div className="list-body">
                    {
                        props.expenses.length === 0 ? (<div className="list-item list-item--message"><span>No expenses!</span></div>) : (props.expenses.map((expense)=>{
                          return <ExpenseListItem key={expense.id} {...expense} /> 
                       }))

                    }
            </div>
          
        {/* spreading the remaining data as ...expense */}

        </div>
    
    );
}

//const ConnectedExpenseList = connect((state)=>{ //first argument is the state from the store, can be given any name
//    return{
//      //name:"nag"
//        expenses:state.expenses
//    };
//})(ExpenseList);
//
//export default ConnectedExpenseList;


//instead of manually assigning to variable, exporting and then importing it, we can have unnamed export like below, this is most prefered way

//instead passing function to Connect as first argument, its common practice to store that function in widely used variable "mapStateToProps" it literally means we are mapping store's State to Props to use it inside component

//when store chnages, the state here(in argument) gets automatically updated and stays upto date with the current data and then re-renders the component when soething chnages, so no need for component to subscribe
const mapStateToProps = (state)=> { //first argument is the state from the store, can be given any name
    return{
      //name:"nag"
        expenses:getVisibleFilters(state.expenses, state.filters),
        //filters:state.filters //we dont want to take data directly from state, because its not filtered or sorted, we want it filtered every time hence sending it to getVisibleFilters component
    };
};

export default connect(mapStateToProps)(ExpenseList);

 



//using it in dashboard file