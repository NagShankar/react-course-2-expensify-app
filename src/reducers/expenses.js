//no dependecy here to import anything because reducers are PURE functions


//Expense Reducer
const expensesReducerDefaultState = [];

const expensesReducer = (state=expensesReducerDefaultState, action) => {
  switch(action.type){
      case 'ADD_EXPENSE' :
          //return state.concat(action.expense), lets use SPEARD syntax like below
          return [
              ...state,
              action.expense
          ]
      case 'REMOVE_EXPENSE':
          return state.filter((expense)=>{
                  return expense.id !== action.id
              })
          
      case 'EDIT_EXPENSE':
          return state.map((expense)=>{
                  if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }                     
                   }else{
                      return expense
                  }
              })
                 
      default:
          return state;
  }  
};


export default expensesReducer;