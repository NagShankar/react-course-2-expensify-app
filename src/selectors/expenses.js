import moment from 'moment';

const getVisibleFilters = ((expenses, {text, sortBy, startDate, endDate}) => {
    
    
/*  

In your filters, startDate and endDate can be undefined. In the case that the startDate or endDate for your filters is undefined, you do not want to filter out any data based on that criteria. If startDate is undefined, typeof startDate !== 'number' will be true, and startDateMatch will be true.

In the case that you provide a number to startDateMatch, that means that you do want to filter by date. That means that the first part of your or (||) condition will be false and you will be checking for a match with the second part of the condition (whether the expense date comes before or after the start/endDate).

In short, undefined start and end dates will have typeof date !== 'number' be true and let all data pass through the filter (assuming the expense meets the other criteria as well!!). Start dates and end dates with values will have this condition be false, and you will filter the expenses based on createdAt.

*/
    
    
    return expenses.filter((expense)=>{
        
//        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;//before comparison was done using numbers, now we need to use actual date
//        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;//before comparison was done using numbers, now we need to use actual date
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;//first checking if start date even exist because we have given user to clear range by clicking "X", else setting true to pass the final triple && condition
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;//first checking if start date even exist because we have given user to clear range by clicking "X", else setting true to pass the final triple && condition
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
                       
        return startDateMatch && endDateMatch && textMatch
    }).sort((a,b) => {
      if(sortBy==="date"){
          return a.createdAt < b.createdAt ? 1 : -1
      }
        else if(sortBy==="amount"){
          return a.amount < b.amount ? 1 : -1
      }
        
    })
});

export default getVisibleFilters;