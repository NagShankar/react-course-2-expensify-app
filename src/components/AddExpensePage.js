import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
//...........after adding middleware and modifying addExpense action creator
//import { addExpense } from "./../actions/expenses";
import { startAddExpense } from "./../actions/expenses";


//previosly this was functional component, now changed to class for testing the unconnected version
export class AddExpensePage extends React.Component{
    
    onSubmission = (expense) => {
        //console.log(expense);
        //props.dispatch(addExpense(expense));
         this.props.submitNow(expense);
         this.props.history.push(
          "/"
        ); /* routing back to dashboard page after submission using history given by react router */
      }
    
    render(){
        
        return(
         <div>
               <h1>Add Expense</h1>
               <ExpenseForm onSubmission={this.onSubmission} />
         </div>
        
        );
    }
    
    
}


const mapDispatchToProps = (dispatch) => {
    return{
        submitNow : (expense) => dispatch(startAddExpense(expense)) //changed addExpense(expense) to startAddExpense(expense), i.e we are dispatching async call here
   }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage); //connect gives access to props which allow us to dispatch action
