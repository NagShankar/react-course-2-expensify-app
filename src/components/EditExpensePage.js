import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "./../actions/expenses";

//previosly this was functional component, now changed to class for testing the unconnected version
export class EditExpensePage extends React.Component{
    
    onSubmission = (expense) => {
          //console.log(expense)
        //alert("hi")
          this.props.editExpense(this.props.expense.id, expense);
          this.props.history.push("/");
        }
    removeNow = (e) => {
         //alert("remove now")
          this.props.removeNow({ id: this.props.expense.id }); //remove expense action accepts "id" as an object, check actions/expenses.js for more details
          this.props.history.push("/"); //redirecting after deleting
        }
       
    render(){
        
         //console.log(this.props); //now props will have the matched expense inside props
        return(
                        
            <div>
           Editing Expense for id: {this.props.match.params.id}
                <ExpenseForm 
                      expense={this.props.expense} /* passing the matched expense as a prop to ExpenseForm component which then can be used to set the default and populate the fields with that data */
                      onSubmission={this.onSubmission}
                />
               <button onClick={this.removeNow}> Remove </button>
          </div>
            
        );
    }
    
}


const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id; //returning the matched expense object from the expenses array
    }),
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeNow: (data) => dispatch(removeExpense( data ))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
