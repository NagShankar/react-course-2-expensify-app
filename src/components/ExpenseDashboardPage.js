import React from "react";
import ExpenseList from './ExpenseList'; //importing connected version of the component, so no need of "{}"
import ExpenseListFilters from './ExpenseListFilters';


const ExpenseDashboardPage = () => (
 <div>
      <p>From dashborad along with Expense list</p>
       <ExpenseListFilters />
       <ExpenseList />
           
    </div>
);


export default ExpenseDashboardPage;