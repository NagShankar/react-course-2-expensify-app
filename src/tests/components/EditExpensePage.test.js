import React from "react";
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses'; 
import { EditExpensePage }  from '../../components/EditExpensePage'; 


//first 3 lines from below test cases are same, hence we can call it only once before each test case with reset or fresh values
let editExpense, removeNow, history, wrapper;

//using one of the Jest's global API tp call once before every ytesy case with fresh data or with reset data 
beforeEach(()=>{
     editExpense = jest.fn();
     removeNow = jest.fn();
     history = { push: jest.fn() }; //history is an object
     wrapper = shallow(<EditExpensePage editExpense={editExpense} history={history} expense={expenses[2]} removeNow={removeNow} match={{params:{id:expenses[2].id}}}/>); 
    
})



//test case EditExpensePage to render correctly
test('should render EditExpensePage correctly',()=>{
   
    expect(wrapper).toMatchSnapshot();
    
})

//test case EditExpensePage to render for edit expense
test('should handle editing expense',()=>{
    wrapper.find('ExpenseForm').prop('onSubmission')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
    
})
    
//test case EditExpensePage to render for remove expense
test('should handle remove expense', ()=>{
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeNow).toHaveBeenLastCalledWith({ id: expenses[2].id });
})