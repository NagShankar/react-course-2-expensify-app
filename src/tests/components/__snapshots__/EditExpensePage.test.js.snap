// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  Editing Expense for id: 
  3
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmission={[Function]}
  />
  <button
    onClick={[Function]}
  >
     Remove 
  </button>
</div>
`;
